include/libvfio-user.h

```diff
@@ -585,26 +585,41 @@ vfu_pci_set_class(vfu_ctx_t *vfu_ctx, uint8_t base, uint8_t sub, uint8_t pi);
 vfu_pci_config_space_t *
 vfu_pci_get_config_space(vfu_ctx_t *vfu_ctx);
 
-/* FIXME does it have to be packed as well? */
-typedef union {
-    struct msicap   msi;
-    struct msixcap  msix;
-    struct pmcap    pm;
-    struct pxcap    px;
-    struct vsc      vsc;
-} vfu_cap_t;
-
-//TODO: Support variable size capabilities.
+#define VFU_CAP_FLAG_EXTENDED (1 << 0)
+#define VFU_CAP_FLAG_READONLY (1 << 1)
 
 /**
- * Setup PCI capabilities.
+ * Add a PCI capability to PCI config space.
+ *
+ * Certain standard capabilities are handled entirely within the library:
+ *
+ * PCI_CAP_ID_EXP (pxcap)
+ * PCI_CAP_ID_MSIX (msixcap)
+ * PCI_CAP_ID_PM (pmcap)
+ *
+ * However, they must still be explicitly initialized and added here.
+ *
+ * The contents of @data are copied in. It must start with either a struct
+ * pci_cap or a struct pci_extcap, with the ID field set (the 'next' field is
+ * ignored). The @flags values allows control over whether a client can write to
+ * the capability. Writes to the "id" or "next" fields are always disallowed.
+ *
+ * If @data is NULL, then it's expected that the region handler for
+ * VFU_PCI_DEV_CFG_REGION_IDX is configured, and can handle access to the
+ * capability.
+ *
+ * Returns the offset of the capability in config space, or -1 on error, with
+ * errno set.
  *
  * @vfu_ctx: the libvfio-user context
- * @caps: array of (vfu_cap_t *)
- * @nr_caps: number of elements in @caps
+ * @cap: capability ID to add
+ * @flags: VFU_CAP_FLAG_*
+ * @data: capability data (including the header), or NULL.
+ * @size: size of the capability including the header
  */
 int
-vfu_pci_setup_caps(vfu_ctx_t *vfu_ctx, vfu_cap_t **caps, int nr_caps);
+vfu_pci_add_capability(vfu_ctx_t *vfu_ctx, uint16_t cap_id,
+                       int flags, void *data, size_t size);
```

Comment:
`size_t size`: We don't need size, the capability itself (even the vendor-specific one) contains enough information to unambiguously determine the size.

Reply:
Not when data == NULL :)

Reply:
removing size